### PR TITLE
Create empty blocks when there is a mismatch in struct schema

### PR DESCRIPTION
**Pull request by bhhari**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by bhhari** [comment]
#14118

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 1]
```diff
@@ -751,4 +766,78 @@ public long getRetainedSizeInBytes()
...
+            implements SelectiveStreamReader
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(MissingFieldStreamReader.class).instanceSize();
+        private final Type outputType;
+        private int[] outputPositions;
+        private int outputPositionCount;
+
+        MissingFieldStreamReader(Optional<Type> type)
+        {
+            this.outputType = type.isPresent() ? type.get() : null;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: Is there any reason for making an instance of MissingFieldStreamReader if output type is missing? I think not. If so, let's change the constructor to take Type (not Optional<Type>) and add requireNotNull.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 2]
```diff
@@ -751,4 +766,78 @@ public long getRetainedSizeInBytes()
...
+        @Override
+        public int read(int offset, int[] positions, int positionCount)
+        {
+            outputPositions = initializeOutputPositions(outputPositions, positions, positionCount);
+            outputPositionCount = positionCount;
+            return outputPositionCount;
+        }
+
+        @Override
+        public int[] getReadPositions()
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: Is this every used? Can it be implemented to throw NotImplementedException? Then, outputPositions variable can be removed and getRetainedSizeInBytes can become empty.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 3]
```diff
@@ -143,34 +147,45 @@ public StructSelectiveStreamReader(
...
+                    .collect(toImmutableMap((f) -> f.getName()
+                            .orElseThrow(() -> new IllegalArgumentException(
+                                    "ROW type does not have field names declared: " + this.outputType))
+                            .toLowerCase(Locale.ENGLISH), Function.identity()))
+                    : ImmutableMap.of();
+            Set<String> structFields = outputType.isPresent() ? nestedTypes.keySet() : nestedStreams.keySet();
+            for (String fieldName : structFields) {
+                StreamDescriptor nestedStream = nestedStreams.get(fieldName);
                 boolean requiredField = requiredFields.map(names -> names.containsKey(fieldName)).orElse(outputRequired);
+                Optional<Type> fieldOutputType = nestedTypes.get(fieldName) != null ? Optional.of(nestedTypes.get(fieldName).getType()) : Optional.empty();
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: use Optional.ofNullable

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 4]
```diff
@@ -143,34 +147,45 @@ public StructSelectiveStreamReader(
...
             this.orderedNestedReaders = new SelectiveStreamReader[0];
         }
         else if (outputRequired || !fieldsWithFilters.isEmpty()) {
             ImmutableMap.Builder<String, SelectiveStreamReader> nestedReaders = ImmutableMap.builder();
-            for (int i = 0; i < nestedStreams.size(); i++) {
-                StreamDescriptor nestedStream = nestedStreams.get(i);
-                String fieldName = nestedStream.getFieldName().toLowerCase(Locale.ENGLISH);
-                Optional<Type> fieldOutputType = nestedTypes.isPresent() ? Optional.of(nestedTypes.get().get(i)) : Optional.empty();
+            Map<String, Field> nestedTypes = outputType.isPresent() ? ((RowType) this.outputType).getFields().stream()
+                    .collect(toImmutableMap((f) -> f.getName()
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: * f -> field
* remove parenthesis around f: (f) -> field

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java (hunk 2 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: This method is called for all the fields, so we will need it

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java (hunk 3 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: the Optional.empty is based on nestedTypes containing the fieldName, if its present then I m retrieving the subfield type. nestesTypes.get(field).geType

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 5]
```diff
@@ -864,6 +872,25 @@ public void testSchemaEvolution()
         assertUpdate("DROP TABLE test_schema_evolution");
     }
 
+    @Test
+    public void testStructSchemaEvolution()
+            throws IOException
+    {
+        assertUpdate("CREATE TABLE struct_test(field0 ROW(a int, b int))");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: Thanks for adding a test. I have some suggestions to make the test a little easier to read.

- CREATE and INSERT queries can be combined using CREATE TABLE AS SELECT (CTAS)
- filePathOld -> oldFilePath
- filePathNew -> newFilePath
- getSinglePath -> getOnlyPath to match Guava naming
- MaterializedResult -> assertQuery

```
        getQueryRunner().execute("CREATE TABLE test_struct(x) AS SELECT CAST(ROW(1, 2) AS ROW(a int, b int)) AS x");
        getQueryRunner().execute("CREATE TABLE test_struct_add_column(x) AS SELECT CAST(ROW(1, 2, 3) AS ROW(a int, b int, c int)) AS x");
        Path oldFilePath = getOnlyPath("test_struct");
        Path newDirectoryPath = getOnlyPath("test_struct_add_column").getParent();
        Files.move(oldFilePath, Paths.get(newDirectoryPath.toString(), "old_file"), ATOMIC_MOVE);

        assertQuery("SELECT * FROM test_struct_add_column", "SELECT (1, 2, 3) UNION ALL SELECT (1, 2, null)");
```

File: presto-orc/src/test/java/com/facebook/presto/orc/TestMapFlatSelectiveStreamReader.java [hunk 6]
```diff
@@ -275,8 +287,8 @@ public void testStructWithNull()
         runTest(
                 "test_flat_map/flat_map_struct_with_null.dwrf",
                 INTEGER,
-                rowType(INTEGER, INTEGER, INTEGER),
-                ExpectedValuesBuilder.get(Function.identity(), TestMapFlatSelectiveStreamReader::intToList).setNullValuesFrequency(SOME));
+                STRUCT_TYPE,
+                ExpectedValuesBuilder.get(Function.identity(), TestMapFlatSelectiveStreamReader::intToList).setNullValuesFrequency(TestMapFlatSelectiveStreamReader.ExpectedValuesBuilder.Frequency.SOME));
```
Comment (mbasmanova) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestMapFlatSelectiveStreamReader.java]: static import TestMapFlatSelectiveStreamReader.ExpectedValuesBuilder.Frequency.SOME

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 7]
```diff
@@ -751,4 +766,79 @@ public long getRetainedSizeInBytes()
...
+        public int[] getReadPositions()
+        {
+            return outputPositions;
+        }
+
+        @Override
+        public Block getBlock(int[] positions, int positionCount)
+        {
+            checkState(outputType != null, "This stream reader produces null block ");
+            return RunLengthEncodedBlock.create(outputType, null, positionCount);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: return createNullBlock(outputType, positionCount);

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 8]
```diff
@@ -751,4 +766,79 @@ public long getRetainedSizeInBytes()
...
+        }
+
+        @Override
+        public BlockLease getBlockView(int[] positions, int positionCount)
+        {
+            checkState(outputType != null, "This stream reader produces null block");
+            return ClosingBlockLease.newLease(RunLengthEncodedBlock.create(outputType, null, positionCount));
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: return ClosingBlockLease.newLease(createNullBlock(outputType, positionCount));

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 9]
```diff
@@ -751,4 +766,79 @@ public long getRetainedSizeInBytes()
...
+        @Override
+        public int[] getReadPositions()
+        {
+            return outputPositions;
+        }
+
+        @Override
+        public Block getBlock(int[] positions, int positionCount)
+        {
+            checkState(outputType != null, "This stream reader produces null block ");
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: This check can never fail. Remove.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 10]
```diff
@@ -751,4 +766,79 @@ public long getRetainedSizeInBytes()
...
+            implements SelectiveStreamReader
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(MissingFieldStreamReader.class).instanceSize();
+        private final Type outputType;
+        private int[] outputPositions;
+        private int outputPositionCount;
+
+        MissingFieldStreamReader(Type type)
+        {
+            requireNonNull(type, "type is required");
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: combine with assignment

`this.outputType = requireNonNull(type, "type is required");`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 11]
```diff
@@ -751,4 +766,79 @@ public long getRetainedSizeInBytes()
...
+        public Block getBlock(int[] positions, int positionCount)
+        {
+            checkState(outputType != null, "This stream reader produces null block ");
+            return RunLengthEncodedBlock.create(outputType, null, positionCount);
+        }
+
+        @Override
+        public BlockLease getBlockView(int[] positions, int positionCount)
+        {
+            checkState(outputType != null, "This stream reader produces null block");
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: This check can never fail. Remove.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java [hunk 12]
```diff
@@ -751,4 +766,79 @@ public long getRetainedSizeInBytes()
...
+        {
+            requireNonNull(type, "type is required");
+            this.outputType = type;
+        }
+
+        @Override
+        public int read(int offset, int[] positions, int positionCount)
+        {
+            outputPositions = initializeOutputPositions(outputPositions, positions, positionCount);
+            outputPositionCount = positionCount;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: outputPositionCount variable is not used; remove it

`return positionCount;`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java (hunk 3 above)
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java]: Optional.ofNullable(nestedTypes.get(fieldName)).map(Field::getType)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 13]
```diff
@@ -1055,6 +1068,12 @@ private Path getPartitionDirectory(String tableName, String partitionClause)
         return Paths.get(filePath).getParent();
     }
 
+
+    private Path getSinglePath(String tableName) {
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: nit: getOnlyPath might be a better name - https://guava.dev/releases/20.0/api/docs/com/google/common/collect/Iterables.html#getOnlyElement-java.lang.Iterable-T-

**Pull request merged**